Solutions/Marain.Services.Tenancy/Marain/Services/Tenancy/Internal/MarainServiceTenancyHelper.cs

```diff
@@ -0,0 +1,83 @@
+﻿// <copyright file="MarainServiceTenancyHelper.cs" company="Endjin Limited">
+// Copyright (c) Endjin Limited. All rights reserved.
+// </copyright>
+
+namespace Marain.Services.Tenancy.Internal
+{
+    using System;
+    using System.Threading.Tasks;
+    using Corvus.Tenancy;
+    using Corvus.Tenancy.Exceptions;
+    using Marain.TenantManagement;
+    using Menes.Exceptions;
+
+    /// <summary>
+    /// Provides methods required by Marain services to validate and work with tenants.
+    /// </summary>
+    public class MarainServiceTenancyHelper : IMarainServiceTenancyHelper
+    {
+        private readonly ITenantProvider tenantProvider;
+        private readonly MarainServiceConfiguration serviceConfiguration;
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="MarainServiceTenancyHelper"/> class.
+        /// </summary>
+        /// <param name="tenantProvider">The tenant management service.</param>
+        /// <param name="serviceConfiguration">Service configuration for the current service.</param>
+        public MarainServiceTenancyHelper(
+            ITenantProvider tenantProvider,
+            MarainServiceConfiguration serviceConfiguration)
+        {
+            this.tenantProvider = tenantProvider;
+            this.serviceConfiguration = serviceConfiguration;
+        }
+
+        /// <inheritdoc/>
+        public async Task<ITenant> GetRequestingTenantAsync(string tenantId)
+        {
+            ITenant tenant = await this.GetTenantAsync(tenantId).ConfigureAwait(false);
+
+            // Validate it's of the expected type. This will throw an ArgumentException if the tenant is not of the expected
+            // type. This is not particularly useful, so we will catch this and instead throw an exception that will result
+            // in a Bad Request response.
+            try
+            {
+                tenant.EnsureTenantIsOfType(MarainTenantType.Client, MarainTenantType.Delegated);
+            }
+            catch (ArgumentException)
+            {
+                throw new OpenApiNotFoundException($"The specified tenant Id, '{tenantId}', is of the wrong type for this request");
```

Comment:
Is this really the right exception to throw? You say you want "Bad Request" but surely that will produce a "Not Found"?

Reply:
Yes. Originally I was going to return BadRequest, but @idg10 and I talked this through and there are a couple of reasons for the resulting change.

Firstly, the existence of anything other than client tenants is an internal implementation detail of the platform. So saying "we found a tenant with the specified ID, but it's not valid for this scenario" is effectively leaking some of our implementation detail.

Following on from this, the path for these requests is generally of the form `/{client tenant Id}/marain/something/whatever`. If your Id doesn't match a client tenant then the response is "there isn't a client tenant with that Id", which better maps onto a 404.

Reply:
Oh. The comment is wrong.